common/persistence/cassandra/cassandraQueue.go

```diff
@@ -165,25 +197,119 @@ func (q *cassandraQueue) DequeueMessages(
 
 	if err := iter.Close(); err != nil {
 		return nil, &workflow.InternalServiceError{
-			Message: fmt.Sprintf("DequeueMessages operation failed. Error: %v", err),
+			Message: fmt.Sprintf("ReadMessages operation failed. Error: %v", err),
 		}
 	}
 
 	return result, nil
 }
 
-func (q *cassandraQueue) Close() error {
-	if q.session != nil {
-		q.session.Close()
+func getMessagePayload(message map[string]interface{}) []byte {
+	return message["message_payload"].([]byte)
+}
+
+func getMessageID(message map[string]interface{}) int {
+	return message["message_id"].(int)
+}
+
+func (q *cassandraQueue) DeleteMessagesBefore(messageID int) error {
+	query := q.session.Query(templateDeleteMessagesQuery, q.queueType, messageID)
+	if err := query.Exec(); err != nil {
+		return &workflow.InternalServiceError{
+			Message: fmt.Sprintf("DeleteMessagesBefore operation failed. Error %v", err),
+		}
 	}
 
 	return nil
 }
 
-func getMessagePayload(message map[string]interface{}) []byte {
-	return message["message_payload"].([]byte)
+func (q *cassandraQueue) insertInitialQueueMetadataRecord() error {
+	version := 0
+	clusterAckLevels := map[string]int{}
+	query := q.session.Query(templateInsertQueueMetadataQuery, q.queueType, clusterAckLevels, version)
+	_, err := query.ScanCAS()
+	if err != nil {
+		return fmt.Errorf("failed to insert initial queue metadata record: %v", err)
+	}
+	// it's ok if the query is not applied, which means that the record exists already.
+	return nil
 }
 
-func getMessageID(message map[string]interface{}) int {
-	return message["message_id"].(int)
+func (q *cassandraQueue) UpdateAckLevel(messageID int, clusterName string) error {
+	queueMetadata, err := q.getQueueMetadata()
+	if err != nil {
+		return &workflow.InternalServiceError{
+			Message: fmt.Sprintf("UpdateAckLevel operation failed. Error %v", err),
+		}
+	}
+
+	queueMetadata.clusterAckLevels[clusterName] = messageID
```

Comment:
to properly handle multiple caller updating the same cluster name -> message ID case, probably should only do update if messageID provided is larger

Reply:
Good catch.